### ISSUE TITLE
Fix thumbnail image generation

### ISSUE THREAD

**Issue by barklund** [swissspidy/web-stories-wp#9749]
## Feature Description

EDIT: A quick fix was implemented meanwhile (#9760), this issue now focuses on the long-term solutions, see also https://github.com/GoogleForCreators/web-stories-wp/issues/9749#issuecomment-992667669

Thumbnail images are generated when a preview page is displayed and there's idle time in the browser.

This happens in the carousel, the checklist, and the grid view.

However, these thumbnail images are generated locally in each component needing them, regardless of whether another preview of the exact same page has been generated elsewhere in the application. Also, this happens even if the thumbnail isn't visible (e.g. in the case of a carousel thumbnail not currently visible due to scroll).

The "generated locally" in the component is extra bad for the new collapsing carousel, as all the pages are unmounted on collapse, which means that on expansion, all the images have to be re-generated.

Furthermore, if multiple page previews are visible at the same time, they might all trigger their idle callback at the same time, and then all the thumbnails will be generated in the same JS frame because there's no queuing or synchronization between the components.

## Solutions considered

Short-term solutions include:

1. Generate images of each page on story load in a queue when the browser is idle (only one image at a time, then request a new idle callback)
  – Then store these images locally in the browser so it's accessible to all components.
  – Update images when a page changes (but in an idle callback).
2. Only generate thumbnail images for the carousel, not in the checklist nor the grid view
3. Only generate thumbnail images for visible pages in the carousel, not the current invisible ones

One or more of the above solutions can be implemented at the same time, last they complement each other rather than compete with each other.

Long-term solutions include:

4. Store images in WP media storage
  – This has some issues with multiple revisions of large stories potentially taking up 100 or even 1000s of physical files on the server
  – Solutions to this include storing only images for the current revision deleting for old revisions as well as storing all images in a single physical file either as a sprite or as a multi-frame image (gif or webp).
5. Store images as base64 in the database along with the story JSON.
  – Base64 wastes about 33% overhead, so images will be even larger than physical files
  – However, the load will be much faster, as only a single load will load both the story data and all the images

For either of the above, image generation scheduling still has to be figured out - when should the image be generated and how do we ensure the browser doesn't hang while it happens? And the above short-term solutions might still all be relevant in the long term.

Both the above have the added advantage of being able to preview stories (as flat images only though, no animation or video) which can be used in both the dashboard and in the Gutenberg blocks inserting stories into pages.

**Comment by barklund** [comment]
A quick experiment on a fairly complicated page with screenshots at 100% size results in these blob/string sizes:

| Type | Quality | Blob | Data URI |
|-|-|-|-|
| image/png | - | 339 kB | 452k characters |
| image/webp | 1.0 | 215 kB | 286k characters |
| image/webp | 0.9 | 31 kB | 42k characters |
| image/webp | 0.7 | 16 kB | 21k characters |
| image/webp | 0.5 | 13 kB | 17k characters |
| image/jpeg | 1.0 | 207 kB | 276k characters |
| image/jpeg | 0.9 | 50 kB | 66k characters |
| image/jpeg | 0.7 | 28 kB | 38k characters |
| image/jpeg | 0.5 | 21 kB | 28k characters |

_Edit_: Feel free to play with [this branch](https://github.com/google/web-stories-wp/tree/try/9749-image-sizes), to try [various settings](https://github.com/google/web-stories-wp/blob/try/9749-image-sizes/packages/story-editor/src/components/eyedropper/index.js#L96-L106).

**Comment by miina** [comment]
> Only generate thumbnail images for the carousel, not in the checklist nor the grid view

This is a quick win that we could already do for 1.15 now and that should fix the issues with the grid view (since the image generation is currently useless anyway as it's only used once, basically a bug even). I'd suggest fixing that part ASAP in a small PR regardless of the long-term solutions.
Additionally, as a short-term fix to unblock the new carousel logic, could we keep the image blobs state outside of the `PagePreview` (for the carousel only) so that the images would not be regenerated (unless changed) but the rest of the logic would stay the same (for now)?

**Comment by dreamofabear** [comment]
I don't think a disk cache makes sense for page thumbnail caching, since I'd expect these thumbnails to get invalidated quite often during editing. A memory cache with a size roughly equal to the number of visible pages in the carousel should be a better fit.

> I'd suggest fixing that part ASAP in a small PR regardless of the long-term solutions.

+1

> since the image generation is currently useless anyway as it's only used once, basically a bug even

I had a similar thought when we shipped thumbnail caching in the first place. Have we done any instrumentation to measure the benefit of our thumbnail caching scheme today (enabled vs. disabled)? We should balance our additional investment in thumbnail caching against the current + expected benefit.

**Comment by swissspidy** [comment]
> I don't think a disk cache makes sense for page thumbnail caching, since I'd expect these thumbnails to get invalidated quite often during editing.

I think the idea there was to only save to disk when saving the story, so that you have a cached image when opening the editor next time.

**Comment by miina** [comment]
> Have we done any instrumentation to measure the benefit of our thumbnail caching scheme today (enabled vs. disabled)?

[Some profiling](https://github.com/google/web-stories-wp/pull/7887#issuecomment-858442969) was done originally for Page navigation / initial rendering to compare the display-as-is vs SVG rendering vs rendering thumbnails before implementing it, but it hasn't been done again with the final implementation.

**Comment by miina** [comment]
Looks like the memory situation has improved after disabling the constant thumbnails generation of the grid view but there are still issues, would be great to move forward with this issue.

@swissspidy As part of this issue is storing the thumbnails then depending on the approach there might be some Infra work involved. Let me know if you have any thoughts on storing the images in the DB vs as images in the WP instance. In case of storing in the DB, do you think it makes sense to store as part of the story or would we need a separate endpoint? Is there still any additional information/testing needed from @barklund first for making a decision?

**Comment by swissspidy** [comment]
To  confirm, this would mean generating images locally and caching client-side during editing, and storing the images server-side upon saving?

Having some images stored on disk is nice when opening an existing story, but there is still a need for some client-side mechanism to seriously reduce memory usage during editing.

> Let me know if you have any thoughts on storing the images in the DB vs as images in the WP instance.

It's probably best for WP pod to create some quick POC for storing the images on disk. DB doesn't sound viable.

> In case of storing in the DB, do you think it makes sense to store as part of the story or would we need a separate endpoint?

That's an implementation detail that shouldn't make any difference.

**Comment by dreamofabear** [comment]
A suggestion for priority/approach:

> However, these thumbnail images are generated locally in each component needing them, regardless of whether another preview of the exact same page has been generated elsewhere in the application.

Let's fix this first e.g. with a simple app-wide memory cache with a story+page ID scheme to reduce duplication generations. Less error prone than disk-first caching and we'll need the plumbing anyways.

Expected benefit: reduce CPU and memory usage.

> if multiple page previews are visible at the same time, they might all trigger their idle callback at the same time

Then, fix this e.g. idle callbacks should check the [remaining idle time](https://developer.mozilla.org/en-US/docs/Web/API/IdleDeadline) is greater than some minimum threshold before doing work. 

Expected benefit: reduce UX jank by avoiding CPU contention between active and background work.

> create some quick POC for storing the images on disk. DB doesn't sound viable.

Let's do this last e.g. move the memory cache from (1) to disk. 

+1 that DB is a bad place for this. `localStorage` seems appropriate but any disk-caching scheme will need to be well-designed to avoid disk usage bloat over time.

Expected benefit: reduce CPU usage and active memory on editor load.

**Comment by miina** [comment]
Moving this issue to Ready for Sprint as Morten brought up it's needed in preparations for #9643 which UX has mentioned to be high priority.

We could implement this in steps as [suggested by Will](https://github.com/GoogleForCreators/web-stories-wp/issues/9749#issuecomment-992667669). (Issue to be split up a bit later)

**Comment by miina** [comment]
@littlemilkstudio I saw you created a few follow-up issues in #10842 that seem to be at least partially related to this issue here.

I was planning this issue for Prometheus for 1.20 as a blocker for #9643 (which potentially will be in 1.21 depending on when the UX / product will finish the work).

Let me know if you see any conflict with this issue and what you had in mind.

The latest idea was to explore [Will's suggestion](https://github.com/GoogleForCreators/web-stories-wp/issues/9749#issuecomment-992667669) and split this issue out as needed into sub-issues. It looks like you already did some centralization in #10842 so perhaps that got partially covered already.

**Comment by maxyinger** [comment]
@miina yea #10822 should just fix this. should hopefully allow for us to remove a lot of this logic too because there's a lot of stuff from the new `PageCanvasProvider` that  already supplies the needed functionality.

You could probably close this and just link it in a comment in #10822 for prior context. Then can escalate #10822 to however you were going to prioritize this ticket.

but yup. the architecture added in #10842 solves the exact problems Will was suggesting. 

Yea for #9643 we should definitely use the `PageCanvasProvider` and basically mimic however the carousel interacts with it.

**Comment by BrittanyIRL** [comment]
Closing in favor of #10822 as this is adding redundancy.
